docs/development.rst

```diff
@@ -2,9 +2,177 @@
     :start-after: sec-begin-development
     :end-before: sec-end-development
 
+.. _writing-adapters:
+
 Writing a model adapter
 =======================
 
+Creating an :class:`~openscm.adapter.Adapter` subclass
+******************************************************
+
+- Create your adapter source file in ``openscm/adapters/``, e.g.
+  ``myadapter.py``
+
+- If your adapter needs additional dependencies add them to the
+  ``requirements_extras`` dictionary in ``setup.py`` (see comment
+  there).
+
+- Subclass the :class:`openscm.adapter.Adapter` class:
+
+  .. code:: python
+
+      from ..adapter import Adapter
+
+      class MyAdapter(Adapter):
+
+- Implement the relevant methods (or just do ``pass`` if you do not
+  need to do anything in the particular method). An adapter should not
+  use parts of OpenSCM other than ``ParameterSet``.
+
+  - The :func:`~openscm.adapter.Adapter._initialize_model` method
+    initializes the adapter and is called only once. It should set the
+    default values of mandatory *model-specific* (not :ref:`standard
+    OpenSCM parameters <standard-parameters>`!) parameters in the
+    :class:`~openscm.core.ParameterSet`. Their :ref:`hierarchical
+    names <parameter-hierarchy>` should start with the model/adapter
+    name.
+
+    .. code:: python
+
+        def _initialize_model(self) -> None:
+            # TODO Initialize the model
+            # TODO Set default parameter values:
+            self._parameters.get_writable_scalar_view(
+                ("MyModel", "Specific Parameter"), ("World",), "Unit"
+            ).set(DEFAULT_VALUE)
+
+  - The :func:`~openscm.adapter.Adapter._initialize_run_parameters`
+    method initializes a particular run. It is called before the
+    adapter is used in any way and at most once before a call to
+    :func:`~openscm.adapter.Adapter.run` or
+    :func:`~openscm.adapter.Adapter.step`.
+
+    .. code:: python
+
+        def _initialize_run_parameters(self) -> None:
+            """
+            TODO Initialize run parameters by reading model parameters
+            from `self._parameters` (see below).
+            """
+
+    You should later use the start and stop time of the run using the
+    :attr:`self._start_time <openscm.adapter.Adapter._start_time>` and
+    :attr:`self._stop_time <openscm.adapter.Adapter._stop_time>`
+    attributes.
+
+  - The :func:`~openscm.adapter.Adapter._initialize_model_input` method
+    initializes the input and model parameters of a particular run. It
+    is also called before the adapter is used in any way and at most
+    once before a call to :func:`~openscm.adapter.Adapter.run` or
+    :func:`~openscm.adapter.Adapter.step`.
+
+    This and the
+    :func:`~openscm.adapter.Adapter._initialize_run_parameters` method
+    are separated for higher efficiency when doing ensemble runs for
+    models that have additional overhead for changing run parameters.
```

Comment:
```suggestion
    models that have additional overhead for changing drivers/scenario setup.
```

I can't think of the right words to use to make the distinction between scenario specific stuff (e.g. emissions) and run specific stuff (e.g. ECS). I'm not sure that 'run parameters' and 'model input' is very clear. @rgieseke @chrisroadmap how would you distinguish between stuff that changes every run and stuff the changes once every, say, 600 runs when doing an ensemble of runs?

Reply:
moved to #22